CoreDNS/SkyDNS provider

istalker2 opened the pull request:
This PR adds support for CoreDNS through its etcd middleware.
Because the middleware is backward compatible with SkyDNS this
commit adds support for SkyDNS as well. In fact, new provider
is available under two names in CLI (coredns and skydns).

All interactions with middleware happen through etcd cluster,
whose location (URIs) is specified via --etcd CLI parameter
by default http://localhost:2379).

The provider translates CoreDNS/DkyDNS SRV records to
A/CNAME + optional TXT endpoints, when reading from etcd and
combines A/CNAME with TXT endpoints back into single SRV record
when writing it back.

Also:
* adds github.com/coreos/etcd package to glide.yaml and vendor folder
  because it is used by the provider
* adds "interfaces" packages and moves Registry and Provider interfaces
  declaration there. This is required in order to use registry in
  provider unit tests. Otherwise there going to be circular imports
  because provider UTests would import registry which imports providers
  back.

mikkeloscar commented:
> adds "interfaces" packages and moves Registry and Provider interfaces
declaration there. This is required in order to use registry in
provider unit tests. Otherwise there going to be circular imports
because provider UTests would import registry which imports providers
back.

This mostly hides the circular dependency, by putting them in the same package. It's still there on a logical level. :)

As the provider doesn't depend on the registry (but the other way around) it should be possible to test it without the registry.

istalker2 commented:
TL;DR; the comment from @mikkeloscar has been addressed, "interfaces" package has gone. Thank you!

The reason why I created interfaces package is to use TXT registry in coredns_test.go. Currently this is impossible due to circular imports. This doesn't discard the fact that registries are tight to providers on logical level, but decouples interface from implementations and breaks circular imports.
The reason, why I wanted TXT registry to be used in tests is because registry by itself creates additional TXT endpoints. Also I wanted the test to be more close to the real code flow so that it use provider Records(), calculate real Plan and append missing TXT records. 

However, this makes it to be not a true unit test as it touches several units at once. Also reorganizing code in the same PR with a new feature is not good even though I believe such reorganization is good. So I removed the interfaces package and changed the test

dshulyak commented:
@istalker2 could you please adjust etcd client to work with tls certificates, as in https://github.com/kubernetes/kubernetes/pull/47049 ?

pigmej commented:
@mikkeloscar could you please take yet another look there? What needs to be addressed to merge that (besides obvious conflict in main.go)? Should it wait for https://github.com/kubernetes-incubator/external-dns/issues/228, because it seems that development there stalled a bit.

@istalker2

istalker2 commented:
@dshulyak done. Please see 3rd commit message for details

bcg62 commented:
+1 for this

hjacobs commented:
What is missing to get this merged?

bcg62 commented:
this still alive?

istalker2 commented:
> this still alive?
I rebased the commit, yet to find someone, who will work on getting this PR merged since, unfortunately, need to focus on another project at the moment

gogeof commented:
Is this finished?
Thanks for your hard work, I also need this feature.

gogeof commented:
@istalker2 Could you add a tutorial of coredns in external-dns/docs/tutorials/ teach us how to use it ?

istalker2 commented:
@gogeof You use it exactly the same way as other externanal-dns providers:
1) Use "coredns" or "skydns" as a provider name for the CLI parameter
2) Deploy etcd database that is going to be the backend for both CoreDNS and this provider
3) Configure CoreDNS to use this etcd etcd instance
4) Pass etcd URLs to the external-dns through the ETCD_URLS environment variable. See commit message for the last of the 3 commits here for a list of supported environment variables

gogeof commented:
@istalker2 thanks a lot
I build from your repo, and run external-dns in my env, but get these message:
```
# docker logs -f 89b90ac28122
time="2017-12-01T05:40:36Z" level=info msg="config: &{Master: KubeConfig: Sources:[service ingress] Namespace:default FQDNTemplate: Compatibility: PublishInternal:false Provider:coredns GoogleProject: DomainFilter:[example.org.] AWSZoneType: AzureConfigFile:/etc/kubernetes/azure.json AzureResourceGroup: CloudflareProxied:false InfobloxGridHost: InfobloxWapiPort:443 InfobloxWapiUsername:admin InfobloxWapiPassword: InfobloxWapiVersion:2.3.1 InfobloxSSLVerify:true InMemoryZones:[] Policy:sync Registry:txt TXTOwnerID:default.example.org TXTPrefix: Interval:1m0s Once:false DryRun:false LogFormat:text MetricsAddress::7979 LogLevel:debug Debug:false}" 
time="2017-12-01T05:40:36Z" level=info msg="Connected to cluster at https://10.96.0.1:443" 
time="2017-12-01T05:40:36Z" level=debug msg="No endpoints could be generated from service default/cheddar" 
time="2017-12-01T05:40:36Z" level=debug msg="No endpoints could be generated from service default/kubernetes" 
time="2017-12-01T05:40:36Z" level=debug msg="No endpoints could be generated from service default/stilton" 
time="2017-12-01T05:40:36Z" level=debug msg="No endpoints could be generated from service default/wensleydale" 
time="2017-12-01T05:40:36Z" level=debug msg="No endpoints could be generated from ingress default/cheese" 
time="2017-12-01T05:41:36Z" level=debug msg="No endpoints could be generated from service default/cheddar" 
time="2017-12-01T05:41:36Z" level=debug msg="No endpoints could be generated from service default/kubernetes" 
time="2017-12-01T05:41:36Z" level=debug msg="No endpoints could be generated from service default/stilton" 
time="2017-12-01T05:41:36Z" level=debug msg="No endpoints could be generated from service default/wensleydale" 
time="2017-12-01T05:41:36Z" level=debug msg="No endpoints could be generated from ingress default/cheese"
```

istalker2 commented:
@gegeof these warnings have nothing to do with the provider code. Actually, it shows, that provider works as expected. The warnings you see are printed by external-dns sources which collect data that is going to be fed to the provider then. external-dns scans various supported k8s objects marked with special annotation and tries to generate a DNS record out of their metadata. There can be several different approaches for generating such record out of an object. But if none of them succeed you get such warning. You're going to get it with any provider

gogeof commented:
@istalker2 yes, you're right, I have found the annotations in code:
source/service.go
```
133 // endpointsFromService extracts the endpoints from a service object
134 func (sc *serviceSource) endpoints(svc *v1.Service) []*endpoint.Endpoint {
135 	var endpoints []*endpoint.Endpoint
136 
137 	// Get the desired hostname of the service from the annotation.
138 	hostnameAnnotation, exists := svc.Annotations[hostnameAnnotationKey]
139 	if !exists {
140 		return nil
141 	}
142 
143 	hostnameList := strings.Split(strings.Replace(hostnameAnnotation, " ", "", -1), ",")
144 	for _, hostname := range hostnameList {
145 		endpoints = append(endpoints, sc.generateEndpoints(svc, hostname)...)
146 	}
147
148	return endpoints
149 }
```
and find the annotation here:
source/source.go
```
32 	hostnameAnnotationKey = "external-dns.alpha.kubernetes.io/hostname"
```

**This is my fault, maybe I should learn more how to use external-dns, I will continue. Thanks a lot.**

gogeof commented:
@istalker2 thanks a lot, I use your repo to build external-dns, and after test, it work for me.

johnmarcou commented:
Hello,

Thank you for this pull request. I am testing external-dns with coredns provider, it works really fine.
I am just experimenting a little issue I would like to report though: ingress rules with multiple IPs addresses create a flapping effect on the DNS resolution, returning sometimes the IP list, sometimes a single IP.

Context:
--------
On our on-prem Kubernetes cluster, the Nginx ingress controller is deployed on two nodes (`10.24.100.2,3`) as entrypoints for the ingress rules. This is a description of one of our ingress rules for example:
```
Name:             kops-kube-ops-view
Namespace:        kube-system
Address:          10.24.100.2,10.24.100.3
Default backend:  default-http-backend:80 (<none>)
TLS:
  SNI routes kops.company.org.au
Rules:
  Host                                Path  Backends
  ----                                ----  --------
  kops.company.org.au
                                      /   kops-kube-ops-view:80 (10.25.7.68:8080)
Annotations:
Events:  <none>
```

When external-dns starts, it creates the 2 DNS records properly:
```
time="2017-12-17T23:44:39Z" level=info msg="Add/set key /skydns/au/org/company/kops/2f7a904e to Host=10.24.100.2, Text="heritage=external-dns,external-dns/owner=default""
time="2017-12-17T23:44:39Z" level=info msg="Add/set key /skydns/au/org/company/kops/6074de14 to Host=10.24.100.3, Text="heritage=external-dns,external-dns/owner=default""
```
You can notice:
 - `2f7a904e = 10.24.100.2`
 - `6074de14 = 10.24.100.3`

```
# nslookup kops.company.org.au 10.24.0.1
Server:		10.24.0.1
Address:	10.24.0.1#1053

Name:	kops.company.org.au
Address: 10.24.100.2
Name:	kops.company.org.au
Address: 10.24.100.3
```

The issue
---------
But on the next updates, after `--interval` time, we are facing two issues:
 - next update apply only on a single `etcd item` / `dns record`
 - the updated record doesn't match with its original IP, which result in a duplication of IP
```
time="2017-12-17T23:44:59Z" level=info msg="Add/set key /skydns/au/org/company/kops/2f7a904e to Host=10.24.100.3, Text="heritage=external-dns,external-dns/owner=default""
```
You can notice:
 - `2f7a904e = 10.24.100.2`
 - `6074de14 = 10.24.100.2`

```
# nslookup kops.company.org.au 10.24.0.1
Server:		10.24.0.1
Address:	10.24.0.1#1053

Name:	kops.company.org.au
Address: 10.24.100.2
```
Across multiple updates from external-dns, DNS server returns the first, the second, or both IPs. 

Expected
---------
The DNS server should always return all the IP addresses.

Thank you for your help.

gogeof commented:
I have check this morning in k8s v1.8.6, it doesn't work. maybe something change.

linki commented:
Merged-in latest master branch

AntonioMeireles commented:
hi all !

what's current status on this ? any ETA / blockers re: getting this merged ?

Best regards!

njuettner commented:
Thanks again for your PR, will be released this week.
